sim/firesim-lib/src/main/cc/bridges/dromajo.cc

```diff
@@ -216,15 +217,19 @@ int dromajo_t::invoke_dromajo(uint8_t* buf) {
 /**
  * Read queue and co-simulate
  */
-void dromajo_t::process_tokens(int num_beats) {
+size_t dromajo_t::process_tokens(int num_beats, size_t minimum_batch_beats) {
+    auto beat_bytes = DMA_DATA_BITS / 8;
+    size_t maximum_batch_bytes = beats * beat_bytes;
+    size_t minimum_batch_bytes = minimum_batch_beats * beat_bytes;
     // TODO: as opt can mmap file and just load directly into it.
-    alignas(4096) uint8_t OUTBUF[QUEUE_DEPTH * sizeof(uint64_t) * 8];
-    pull(this->_dma_addr, (char*)OUTBUF, num_beats * sizeof(uint64_t) * 8);
+    alignas(4096) uint8_t OUTBUF[maximum_batch_bytes];
+    auto bytes_received = pull(stream_idx, (char*)OUTBUF, maximum_batch_bytes, minimum_batch_bytes);
 
     // skip if co-sim not enabled
-    if (!this->dromajo_cosim) return;
+    if (!this->dromajo_cosim)
+        return bytes_received;
```

Comment:
As you figured, the right play is 100% to drop the tokens in the bridge module. This is what tracerv does (it didn't always). I really didn't want to change this bridge too much because i have no idea how to test it and there's nothing in CI for it. @abejgonzalez you could consider doing this in the future. 

Anyways, I'll make it return 0. I was consistently making these process tokens functions return non-zero if they managed to pull something out of their stream (i.e, from a token perspective, they are making forward progress).